File: examples/src/main/java/net/goui/phonenumbers/examples/ExampleClassifier.java
/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2023, David Beaumont (https://github.com/hagbard).

This program and the accompanying materials are made available under the terms of the
Eclipse Public License v. 2.0 available at https://www.eclipse.org/legal/epl-2.0, or the
Apache License, Version 2.0 available at https://www.apache.org/licenses/LICENSE-2.0.

SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

package net.goui.phonenumbers.examples;

import static java.util.function.Function.identity;
import static net.goui.phonenumbers.FormatType.INTERNATIONAL;
import static net.goui.phonenumbers.FormatType.NATIONAL;

import com.google.common.annotations.VisibleForTesting;
import net.goui.phonenumbers.AbstractPhoneNumberClassifier;
import net.goui.phonenumbers.PhoneNumberFormatter;
import net.goui.phonenumbers.PhoneNumberParser;
import net.goui.phonenumbers.metadata.RawClassifier;

public final class ExampleClassifier extends AbstractPhoneNumberClassifier {
  private static final ExampleClassifier INSTANCE =
      new ExampleClassifier(
          AbstractPhoneNumberClassifier.loadRawClassifier(
              SchemaVersion.of("goui.net/phonenumbers/examples/simple/dfa/minimal", 1)));

  public static ExampleClassifier getInstance() {
    return INSTANCE;
  }

  private final PhoneNumberFormatter nationalFormatter = createFormatter(NATIONAL);
  private final PhoneNumberFormatter internationalFormatter = createFormatter(INTERNATIONAL);
  private final PhoneNumberParser<String> parser = createParser(identity());

  private ExampleClassifier(RawClassifier rawClassifier) {
    super(rawClassifier);
  }

  @VisibleForTesting
  RawClassifier rawClassifierForTesting() {
    return rawClassifier();
  }

  public PhoneNumberFormatter national() {
    return nationalFormatter;
  }

  public PhoneNumberFormatter international() {
    return internationalFormatter;
  }

  public PhoneNumberParser<String> getParser() {
    return parser;
  }
}

File: javascript/tests/example-classifier.ts
/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 Copyright (c) 2023, David Beaumont (https://github.com/hagbard).

 This program and the accompanying materials are made available under the terms of the
 Eclipse Public License v. 2.0 available at https://www.eclipse.org/legal/epl-2.0, or the
 Apache License, Version 2.0 available at https://www.apache.org/licenses/LICENSE-2.0.

 SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

import {
    SchemaVersion,
    PhoneNumber,
    AbstractPhoneNumberClassifier,
    Matcher,
    SingleValuedMatcher,
    DigitSequence,
    PhoneNumberFormatter,
    FormatType,
    PhoneNumberParser } from "../dist/index.js";
import { RawClassifier, Converter } from "../dist/internal.js";
const fs = require("fs");

export enum LpnType {
  PREMIUM_RATE,
  TOLL_FREE,
  SHARED_COST,
  FIXED_LINE,
  MOBILE,
  FIXED_LINE_OR_MOBILE,
  PAGER,
  PERSONAL_NUMBER,
  VOIP,
  UAN,
  VOICEMAIL,
}

export class ExampleClassifier extends AbstractPhoneNumberClassifier {
  private readonly typeMatcher: SingleValuedMatcher<LpnType>;
  private readonly regionMatcher: Matcher<string>;
  private readonly nationalFormatter: PhoneNumberFormatter;
  private readonly internationalFormatter: PhoneNumberFormatter;
  private readonly parser: PhoneNumberParser<string>;

  constructor(path: string) {
    super(
        fs.readFileSync(path, { encoding: "utf8", flag: "r" }),
        new SchemaVersion("goui.net/libphonenumber/dfa/compact", 1));
    this.typeMatcher =
        super.forValues("LPN:TYPE", super.ofNumericEnum(LpnType)).singleValuedMatcher();
    // We don't have a region code enum to hand, so use strings directly.
    this.regionMatcher = super.forStrings("REGION").matcher();
    this.nationalFormatter = super.createFormatter(FormatType.National);
    this.internationalFormatter = super.createFormatter(FormatType.International);
    this.parser = super.createParser(Converter.identity());
  }

  rawClassifierForTests(): RawClassifier {
    return super.getRawClassifier();
  }

  forType(): SingleValuedMatcher<LpnType> { return this.typeMatcher; }

  forRegion(): Matcher<string> { return this.regionMatcher; }

  formatNational(number: PhoneNumber): string {
    return this.nationalFormatter.format(number);
  }

  formatInternational(number: PhoneNumber): string {
    return this.internationalFormatter.format(number);
  }

  getParser(): PhoneNumberParser<string> {
    return this.parser;
  }
}
